idaes/apps/caprese/examples/cstr_mhe.py

```diff
@@ -0,0 +1,229 @@
+
+__author__ = "Kuan-Han Lin"
+
+
+
+# See if ipopt is available and set up solver
+if SolverFactory('ipopt').available():
+    solver = SolverFactory('ipopt')
+    solver.options = {
+            'tol': 1e-6,
+            'bound_push': 1e-8,
+            'halt_on_ampl_error': 'yes',
+            'linear_solver': 'ma57',
+            }
+else:
+    solver = None
+
+
+def main():
+    m_estimator = make_model(horizon=8, ntfe=16, ntcp=2, bounds=True)
+    sample_time = 0.5
+    m_plant = make_model(horizon=sample_time, ntfe=5, ntcp=2)
+    time_plant = m_plant.fs.time
+
+    simulation_horizon = 20
+    n_samples_to_simulate = round(simulation_horizon/sample_time)
+
+    samples_to_simulate = [time_plant.first() + i*sample_time
+                           for i in range(1, n_samples_to_simulate)]
+
+    # We must identify for the estimator which variables are our
+    # inputs and measurements.
+    inputs = [
+            m_plant.fs.mixer.S_inlet.flow_vol[0],
+            m_plant.fs.mixer.E_inlet.flow_vol[0],
+            ]
+    measurements = [
+            m_plant.fs.cstr.outlet.conc_mol[0, 'C'],
+            m_plant.fs.cstr.outlet.conc_mol[0, 'E'],
+            m_plant.fs.cstr.outlet.conc_mol[0, 'S'],
+            m_plant.fs.cstr.outlet.conc_mol[0, 'P'],
+            m_plant.fs.cstr.outlet.temperature[0],
+            m_plant.fs.cstr.volume[0],
+            ]
+    
+    # Construct the "MHE simulator" object
+    mhe = DynamicSim(
+            plant_model=m_plant,
+            plant_time_set=m_plant.fs.time,
+            estimator_model=m_estimator, 
+            estimator_time_set=m_estimator.fs.time,
+            inputs_at_t0=inputs,
+            measurements_at_t0=measurements,
+            sample_time=sample_time,
+            )
+    
+    plant = mhe.plant
+    estimator = mhe.estimator
+    
+    p_t0 = mhe.plant.time.first()
+    c_t0 = mhe.estimator.time.first()
+    p_ts = mhe.plant.sample_points[1]
+    c_ts = mhe.estimator.sample_points[1]
+    #--------------------------------------------------------------------------
+    # Declare variables of interest for plotting.
+    # It's ok not declaring anything. The data manager will still save some 
+    # important data, but the user should use the default string of CUID for plotting afterward.
+    states_of_interest = (Reference(mhe.plant.mod.fs.cstr.control_volume.material_holdup[:,'aq','S']),
+                          Reference(mhe.plant.mod.fs.cstr.control_volume.material_holdup[:,'aq','E']),
+                          # Reference(mhe.plant.mod.fs.cstr.control_volume.material_holdup[:,'aq','C']),
+                          # Reference(mhe.plant.mod.fs.cstr.control_volume.material_holdup[:,'aq','P']),
+                          # Reference(mhe.plant.mod.fs.cstr.control_volume.material_holdup[:,'aq','Solvent']),
+                          Reference(mhe.plant.mod.fs.cstr.control_volume.energy_holdup[:,'aq']),
+                          )
```

Comment:
A little odd that we require the user to specify variables different ways. What I mean is that in `DynamicSim`, we require `inputs_at_t0`, but here we require the time-indexed reference. I think I favor making the user supply either a time-indexed variable (possibly a reference) or a CUID of one. Probably not the most pressing thing to change, however.